### DISCUSSION TITLE
Support for sprites?

### DISCUSSION THREAD

**Discussion by PlastiBots** [moononournation/Arduino_GFX#243] [General]
Per the  subject, does this library support sprites? Thx.

**Comment by moononournation** [comment]
it should be something like canvas. But since I do not have experience on using sprites, the feature may not enough for actual use. Can you give me your requirements?

**Reply by h5n1xp** [reply]
I think it should be quite easy to implement sprites using canvas objects if the class exposed its underlying image data. One could then "blit" that image data to another (framebuffer) canvas/display object using a stencilling method of choice.

**Comment by PlastiBots** [comment]
I can have a look at the canvas feature.  For Sprites, the general idea is to have smooth overlay graphics that can be animated/moved without a complete refresh of the screen.  At least that's the way I understand it.  Here's a good example that helps describe it:  https://www.youtube.com/watch?v=sRGXQg7028A

**Comment by moononournation** [comment]
It should be the job of higher presentation layer library like LVGL.

**Reply by PlastiBots** [reply]
Maybe, but in the case of T-RGB, I don't want to use LVGL, and I see your library works with that device (which is great BTW - so thx!).   TFT-eSPI was an option as it has sprits, but it doesn't (yet) support that display type.

**Reply by moononournation** [reply]
Your contribution on this scope is welcome, but it should be a long way to go.

**Comment by h5n1xp** [comment]
I have written a simple sprite class which subclasses the canvas type (happy to share the code), but in order to implement transparency with any speed the canvas class needs to publicly expose its _framebuffer pointer.

**Comment by moononournation** [comment]
Ok, exposed. Please get update to test your class.

**Reply by h5n1xp** [reply]
The getFramebuffer() method is great, and I can get the sprite drawing methods to run as fast as possible now.

@PlastiBots i guess we just need to discuss what you need from the API.

I have implemented what I need (Commodre Amiga style ChromaKey with backing store).

-edit- I am also wondering about using DMA (and/or the second core in the ESP32-S3) to improve performance.

**Reply by PlastiBots** [reply]
Thanks for this.  Apologies as I have limited time at this moment to focus on this (daytime job keeping me very busy) :-)    Ideally, what I am looking for is to have the ability for smooth scrolling of graphics/images using sprites as in other libraries (e.g. TFT_eSPI).

**Reply by h5n1xp** [reply]
Well, for smooth. scrolling, what we might need to think about is a canvas larger than the display and allowing a display offset, so  when the canvas is drawn to the display during the flush() operation , the displayable  area is moved around (again that's how smooth scrolling was implemented back in the Amiga days).

**Reply by h5n1xp** [reply]
@moononournation I've been reading through the ADAFruit_GFX code and notice that they expose the canvas frameBuffer with the function: 

uint16_t* getBuffer(void)

Not sure how compatible you want Arduino_GFX to be with the ADAFruit version, but it might make sense to keep the APIs as close as possible.

**Reply by h5n1xp** [reply]
Also, the Adafruit_GFX library allocates its buffers in the constructor, but Arduino_GFX allocates it's buffers in the begin() method is there a reason for this difference?

It makes more sense to me that the buffers would be allocated in the constructor, and the begin() method would just be a NOP on the canvas class since it only really makes sense on the underlying (direct) display class.

**Reply by moononournation** [reply]
Actually I don't know Adafruit implemented canvas.

**Reply by h5n1xp** [reply]
It might be a new addition to the Adafruit_gfx library? It's defined in the main c++ file. Possibly they have copied you, as a Canvas class is a really useful thing to have! 

I'm thinking that with a few additions to your canvas class it could be a very powerful graphic primitive!

**Reply by h5n1xp** [reply]
Also, I wonder if the canvas performance could be improved by allocating its buffer in main ram (rather than PSRAM), perhaps a flag in the constructor to select the ram allocation type?

Just tested, and confirmed that the canvas is faster when allocated in main ram (only a little bit)

**Comment by h5n1xp** [comment]
> Per the  subject, does this library support sprites? Thx.  

Ok, a few evenings of work and I have a little tech demo working.

This is built by subclassing the canvas class.

Currently runs about 23FPS... which is quite impressive given the amount of data being shifted each frame! But I wonder if I can get it to run faster.

https://youtu.be/r1OnUF03_B4[Smooth scrolling Tech Demo](https://youtu.be/r1OnUF03_B4)

**Reply by moononournation** [reply]
looks very good, any feature I can enhance to help this?

**Reply by h5n1xp** [reply]
I would like to share my tech demo code with you privately, so you can perhaps make the API more consistent with your classes, and maybe you could look for optimisations I missed? 

I can't share publicly because I don't own the graphics assets, I will make some public domain graphics assets in the future so I can share the demo code.

I raised an issue regarding the indexed Canvas, I'm not sure how much performance improvement that would bring as I am still testing and don't have much time.

**Reply by h5n1xp** [reply]
Also, I've been looking through the TFT_eSPI code and trying to see how/if the DMA is being used on the 8bit ESP32-S3 LCD interface, and could such a thing be used on Arduino_GFX for uploading the canvas framebuffer to the display?

**Comment by moononournation** [comment]
You can replace with free assets before expose it to public.

**Comment by h5n1xp** [comment]
An updated video, the Sprite library now supports palette based images to halve RAM usage (each image can have its own palette and expanded to 16bit upon drawing to the Canvas).

The palette based Blitter copy is a bit exotic as 8bit RAM reads are very slow on the ESP32, so 4pixels are fetched as single 32bit read, then unpacked in CPU registers during the blit loop.

https://youtu.be/wZR8tQ5Uga4

**Reply by h5n1xp** [reply]
Over the weekend, I will put up an alpha of my Sprite class on my GitHub and post a link here, perhaps @PlastiBots could test and see what he thinks?

**Reply by h5n1xp** [reply]
As Promised, some code to test.

https://github.com/h5n1xp/Arduino_Sprite

**Comment by moononournation** [comment]
how you calculate FPS before?

**Reply by h5n1xp** [reply]
A timer thread (pinned to core 0 on the ESP32) which increments a variable every second, then divide the total number of frames which have been generated by the number of seconds which passed. 

I didn't include this in my pubished demo as it is ESP32/FreeRTOS specific:


float FPS = 0;
float secondsTick = 0;
float frames = 0;

void SecondTickerCode(void* p){

  while(1){
    vTaskDelay(1000/portTICK_PERIOD_MS); //sleep thread for 1 second 
    secondsTick +=1;
    FPS = frames / secondsTick;
  }
}

void setup() {
     xTaskCreatePinnedToCore(SecondTickerCode,"Seconds Timer",1000,NULL,-1,&SecondTicker,0);
}

void loop(void) {
  frames +=1;
  gfx->flush();
}

**Reply by h5n1xp** [reply]
I have a more extreme video now:

https://youtu.be/56ELuOAAgyg

**Comment by h5n1xp** [comment]
Ok, I've been playing around with my sprite library, and have decided that subclassing from the canvas class probably isn't the best way to go.

Instead, I will use composition, and have the sprite class allocate its own canvas internally as needed. The internal canvas object won't be private or protected so drawing operations can be exposed if needed.

I'm implementing a more structured scene class to allow multiple sprites to be managed easily and cleanly. So this makes more sense, despite some. Serious API breakage, hopefully no one is actually using my class yet :-)

**Reply by moononournation** [reply]
Good to hear that, keep fighting💪🏻

**Reply by h5n1xp** [reply]
Took ages to find some time, but I have uploaded the latest code to my GitHub. This includes the more intensive demo, with 10(?) levels of parallax.

The architecture is now split into two parts, the Sprite class which handles all the high level details of managing sprites (including animation support), and a Blitter class which handles the low level pixel data (users shouldn't need to use this), this way the single sprite class can decide how to handle the different image formats.

This needs the latest Arduino_GFX library as it depends upon the new features like GetFramebuffer() and the GFX_SKIP_OUTPUT_BEGIN define.

**Comment by h5n1xp** [comment]
Apologies for the lack of updates. I have been working with @PlastiBots and we now have a working sprite library. 
My main focus is 2D games, and @PlastiBots has been focusing on using it for building richly animated Graphical User Interfaces!

Currently I consider the library in Beta. Please test with the Arduino_GFX library on as many different hardware configurations as possible, and post issues/bugs/feature requests on my Github.

https://github.com/h5n1xp/Arduino_Sprite

**Reply by moononournation** [reply]
I would like to make some fine tunes in GFX for your project, could you please expose the version of example that showing FPS for easy reference?

**Reply by h5n1xp** [reply]
Certainly, I'll try and have a version for you online this morning! Many thanks.

**Reply by h5n1xp** [reply]
> I would like to make some fine tunes in GFX for your project, could you please expose the version of example that showing FPS for easy reference?

Ok, I have uploaded a version of the example, with an FPS counter in the top left corner of the screen here:

https://github.com/h5n1xp/Arduino_Sprite/tree/main/src/testing

-edit- Let me know if you have any issues and I will fix them ASAP.

**Reply by moononournation** [reply]
ok, take a baseline video before tuning for reference: https://youtu.be/5JN2xYd2sbc

**Reply by h5n1xp** [reply]
> ok, take a baseline video before tuning for reference: https://youtu.be/5JN2xYd2sbc

Are you sure you aren't using a better version of Arduino_GFX than I have? You are getting an excellent ~30FPS, but I get about 25FPS here, I think on the same Lilygo T3 board!

https://youtube.com/shorts/w9kNDNZN0kU?feature=share

**Reply by moononournation** [reply]
Yes, I am tuning on parallel 8-bit interface. But I found tuned ESP32LCD8 not compatible with T-Display-S3, so I am using ESP32PAR8Q

**Reply by h5n1xp** [reply]
> Yes, I am tuning on parallel 8-bit interface. But I found tuned ESP32LCD8 not compatible with T-Display-S3, so I am using ESP32PAR8Q

So it looks like (Tuned) ESP32PAR8Q is faster than ESP32LCD8 which I have been using for the T-Display-S3? 🤔

**Reply by moononournation** [reply]
actually ESP32LCD8 is much faster than before, but it is not compatible to T-Display-S3

**Reply by h5n1xp** [reply]
> actually ESP32LCD8 is much faster than before, but it is not compatible to T-Display-S3

Do you have an idea what the specific issue with the T-Display-S3 might be?

**Reply by h5n1xp** [reply]
> actually ESP32LCD8 is much faster than before, but it is not compatible to T-Display-S3

Also, I want to play with your new version of the Arduino_GFX library!! 😁

**Reply by moononournation** [reply]
the version in video already at master branch

**Reply by h5n1xp** [reply]
> the version in video already at master branch

I can confirm, using the master branch, that ESP32PAR8Q does give me approximately 30FPS! That's amazing! This is wonderful!


Note:
I tested the ESP32LCD8 bus driver, but yes, I get corrupted gfx at this time.

ESP32PAR8 gives 20FPS.
ESP32PAR8QQ just gives black screen.

**Reply by h5n1xp** [reply]
Ok more testing using the current Master Branch, this time with my Lilygo T-HMI (Essentially a 320x240 version of the T-Display-S3), using ESP32PAR8Q and setting a display size of 320x170. The test demo also gives 30FPS (as expected, since the CPU/Display hardware is identical). 

But setting display size to 320x200, drops frame rate to 25FPS.
Then setting display size to 320x240 (full screen), the frame rate drops to 21FPS.

**Reply by moononournation** [reply]
Yes, the flush() speed very depends on the resolution. Can you also check the 320x240 screen can work with ESP32LCD8?

**Reply by h5n1xp** [reply]
> Yes, the flush() speed very depends on the resolution. Can you also check the 320x240 screen can work with ESP32LCD8?

The T-HMI with display size of 320x240 using ESP32LCD8 doesn't work, just a white screen (the T-HMI uses inverted display colours).

**Reply by moononournation** [reply]
the pin mapping is exactly the same?

**Reply by h5n1xp** [reply]
Yes, I just comment out whichever bus driver I don't want to test:

`Arduino_DataBus *bus = new Arduino_ESP32LCD8(PIN_LCD_DC, PIN_LCD_CS, PIN_LCD_WR, PIN_LCD_RD, PIN_LCD_D0, PIN_LCD_D1, PIN_LCD_D2, PIN_LCD_D3, PIN_LCD_D4, PIN_LCD_D5, PIN_LCD_D6, PIN_LCD_D7);

 Arduino_DataBus *bus = new Arduino_ESP32PAR8Q(PIN_LCD_DC, PIN_LCD_CS, PIN_LCD_WR, PIN_LCD_RD, PIN_LCD_D0, PIN_LCD_D1, PIN_LCD_D2, PIN_LCD_D3, PIN_LCD_D4, PIN_LCD_D5, PIN_LCD_D6, PIN_LCD_D7);`

**Reply by moononournation** [reply]
A progress update for using Canvas_Indexed:
https://youtu.be/F5VOuaRSVO8

It should use a unified color index. And the animation image sequence align vertical should be much effective.

**Comment by moononournation** [comment]
Your code is already tuned speed very well, seems not too much can tune: 
https://youtu.be/_MR483mlOHk
But I have refactored some common API to GFX, so it can be much easier to implement and enhance.

**Comment by moononournation** [comment]
And this is the result with same screen but connect with SPI@80MHz:
https://youtu.be/_MR483mlOHk

**Reply by h5n1xp** [reply]
> And this is the result with same screen but connect with SPI@80MHz:
> https://youtu.be/_MR483mlOHk

I just tested ESP32PAR8Q using begin(80000000) which works but offers no performance improvement, perhaps the PSRAM performance is the limit here?

I modified your Arduino_Canvas class (in my local copy) to allocate RAM on the heap rather than the PSRAM, and get 31FPS, but uses a huge amount of RAM.

**Reply by moononournation** [reply]
ESP32PAR8Q is a software class and cannot set the speed.

And here are my simplified sprite class demo source code:
https://github.com/moononournation/Arduino_Sprite/tree/main/src/Sprite_demo
The are 3 type of Sprite demo:
- USE_INDEXED_SPRITE
- USE_INDEXED_SPRITE_AND_CANVAS
- USE_SPRITE

The INDEXED_SPRITE demo refactor the draw function to canvas and the result is similar to your sample with a little bit overhead😅
The INDEXED_SPRITE_AND_CANVAS demo use Canvas_Indexed class and all sprite share same color index. It can potential faster.

**Reply by moononournation** [reply]
The latest version of INDEXED_SPRITE_AND_CANVAS can have 31 FPS:
https://youtu.be/Nx6XlbqKgNU

**Reply by moononournation** [reply]
I think I will use a sprite master image like vintage console game. So I will try to build a helper fuction for manipulating this.

**Reply by h5n1xp** [reply]
> ESP32PAR8Q is a software class and cannot set the speed.
> 
> 
> 
> And here are my simplified sprite class demo source code:
> 
> https://github.com/moononournation/Arduino_Sprite/tree/main/src/Sprite_demo
> 

In this example you use the ESP32PAR8Q bus driver, but also set the Canvas->begin(80000000), is that just left over from SPI or other bus driver testing?

**Reply by moononournation** [reply]
The software interface simply ignore this parameter, it just for SPI

**Reply by h5n1xp** [reply]
> The software interface simply ignore this parameter, it just for SPI

That makes sense, thank you.

Have you managed to get ESP32LCD8 bus driver to work? 

I would really love to get 30FPS on the T-HMI 320x240! 🙂

Let me know what I can do to help!

**Comment by moononournation** [comment]
The first implementation using GIF as a Sprite master source:
https://youtube.com/shorts/9bkRixPYlbo?feature=share

**Reply by h5n1xp** [reply]
This is great, so you are using a gif as the image source? 

That definitely makes loading animations easier 😆

**Reply by moononournation** [reply]
yes, still developing how to implement animation

**Reply by moononournation** [reply]
You can find the WIP sample at:
https://github.com/moononournation/Arduino_Sprite/tree/main/src/GIF_Sprite_demo

**Reply by moononournation** [reply]
https://youtube.com/shorts/DgsqMd7PXTE?feature=share

**Reply by moononournation** [reply]
Now, it is one of Arduino_GFX example:

https://youtube.com/shorts/ZkC1_uSAlVc

**Reply by h5n1xp** [reply]
> Now, it is one of Arduino_GFX example:
> 
> https://youtube.com/shorts/ZkC1_uSAlVc

The performance is exceptional

Is this display connected via the SPI? If so, I'm very impressed how you have tweaked the flush() method!

**Reply by moononournation** [reply]
just because this display can run SPI@80MHz

**Comment by PlastiBots** [comment]
Hey guys.. Pulled together a video of my NeoLmn8 project using a T-RGB Arduino GFX and Matt's Sprite library. Shows off some cool use cases: https://youtu.be/B_X_v7OUKqw
